libntxm/arm7/source/player.cpp

```diff
@@ -149,6 +149,99 @@ void Player::stop(void)
 	resetPanning();
 }
 
+void Player::calcCursorPos(u32 n_ticks)
+{
+	if (!state.playing_piano_sample)
+		return;
+
+	const u8 looptype = state.piano_sample_looptype;
+
+	// promoted and shifted up for fixed point calcs, 
+	// for accuracy. only shift down when passing to sampledisp
+	const u64 loopstart = (u64)state.piano_sample_loopstart << 32;
+	const u64 looplen = (u64)state.piano_sample_looplen << 32;
+	const u64 nsamps = (u64)state.piano_sample_nsamps_total << 32;
+	const u64 loopend = loopstart + looplen;
+	
+	bool& looprev = state.piano_sample_loopreverse;
+	u64& playpos = state.piano_sample_nsamps_position; // alias the name for cleanliness
+
+	if (n_ticks == 0) return;
+	
+	const u64 step = ((u64)state.piano_sample_playfreq << 32) / (1000 / (u64)n_ticks); // timer
+
+	if (looprev) {
+		if (step > playpos) {
+			playpos = loopstart + step - playpos;
+			looprev = false;
+		} else
+			playpos -= step;
+	} else
+		playpos += step;
+
+	if (looptype == FORWARD_LOOP && playpos >= loopend)
+		playpos = loopstart + playpos - loopend;
+	else if (looptype == PING_PONG_LOOP) {
+		if (looprev && playpos <= loopstart) {
+			// 2*playback pos, as the second one is used to calculate
+			// the error that needs to be corrected
+			playpos = 2 * loopstart - playpos;
+			looprev = false;
+		}
+		else if (!looprev && playpos >= loopend) {
+			// same here
+			playpos = 2 * loopend - playpos;
+			looprev = true;
+		}
+	}
+	else if (looptype == NO_LOOP && playpos >= nsamps) {
+		CommandStopCursor();
+		return;
+	}
+
+	if (looptype != NO_LOOP && looplen <= step && playpos > loopstart)
```

Comment:
special case for when `step` is lesser than `looplen` - it will not be able to calculate a position to draw otherwise as it would skip over the loop section